community/kernel/src/main/java/org/neo4j/unsafe/batchinsert/BatchInserterImpl.java

```diff
@@ -342,17 +347,27 @@ public boolean relationshipHasProperty( long relationship, String propertyName )
     @Override
     public void setNodeProperty( long node, String propertyName, Object newValue )
     {
-        propertyCreator.setPrimitiveProperty( getNodeRecord( node ), getOrCreatePropertyKeyId( propertyName ),
-                newValue, recordAccess.getPropertyRecords() );
-        recordAccess.commit();
+        RecordProxy<Long,NodeRecord,Void> nodeRecord = getNodeRecord( node );
+        int propertyKey = getOrCreatePropertyKeyId( propertyName );
+        RecordAccess<Long,PropertyRecord,PrimitiveRecord> propertyRecords = recordAccess.getPropertyRecords();
+
+        removePropertyIfExist( nodeRecord, propertyKey, propertyRecords );
+        propertyCreator.setPrimitiveProperty( nodeRecord, propertyKey, newValue, propertyRecords );
+
+        flushStrategy.flush();
     }
 
     @Override
     public void setRelationshipProperty( long relationship, String propertyName, Object propertyValue )
     {
-        propertyCreator.setPrimitiveProperty( getRelationshipRecord( relationship ),
-                getOrCreatePropertyKeyId( propertyName ), propertyValue, recordAccess.getPropertyRecords() );
-        recordAccess.commit();
+        RecordProxy<Long,RelationshipRecord,Void> relationshipRecord = getRelationshipRecord( relationship );
+        int propertyKey = getOrCreatePropertyKeyId( propertyName );
+        RecordAccess<Long,PropertyRecord,PrimitiveRecord> propertyRecords = recordAccess.getPropertyRecords();
+
+        removePropertyIfExist( relationshipRecord, propertyKey, propertyRecords );
```

Comment:
Setting the property this way and given the changes in `PropertyCreator` makes me think that the `PropertyCreator#setPrimitiveProperty` is completely unnecessary. So instead just use `primitiveAddProperty` and remove `setPrimitiveProperty`.

The drawback of your changes is that the property chain needs to be traversed potentially two times, first to find-and-delete and then to find space to allocate the new one. The purpose of `PropertyCreator#setPrimitiveProperty` was to do that in a single method, with only one chain traversal.